core/coreapi/object.go

```diff
@@ -0,0 +1,190 @@
+package coreapi
+
+import (
+	"bytes"
+	"context"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/ipfs/go-ipfs/merkledag/utils"
+
+	coreiface "github.com/ipfs/go-ipfs/core/coreapi/interface"
+	dag "github.com/ipfs/go-ipfs/merkledag"
+	ft "github.com/ipfs/go-ipfs/unixfs"
+)
+
+type ObjectAPI CoreAPI
+
+func (api *ObjectAPI) New(ctx context.Context) (coreiface.Node, error) {
+	node := new(dag.ProtoNode)
+
+	_, err := api.node.DAG.Add(node)
+	if err != nil {
+		return nil, err
+	}
+	return node, nil
+}
+
+func (api *ObjectAPI) Put(context.Context, coreiface.Node) error {
+	return errors.New("todo") // TODO: what should this method take? Should we just redir to dag-put?f
+}
+
+func (api *ObjectAPI) Get(ctx context.Context, path coreiface.Path) (coreiface.Node, error) {
+	return api.core().ResolveNode(ctx, path)
+}
+
+func (api *ObjectAPI) Data(ctx context.Context, path coreiface.Path) (io.Reader, error) {
+	nd, err := api.core().ResolveNode(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	pbnd, ok := nd.(*dag.ProtoNode)
+	if !ok {
+		return nil, dag.ErrNotProtobuf
+	}
+
+	return bytes.NewReader(pbnd.Data()), nil
+}
+
+func (api *ObjectAPI) Links(ctx context.Context, path coreiface.Path) ([]*coreiface.Link, error) {
+	nd, err := api.core().ResolveNode(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	links := nd.Links()
+	out := make([]*coreiface.Link, len(links))
+	for n, l := range links {
+		out[n] = (*coreiface.Link)(l)
+	}
+
+	return out, nil
+}
+
+func (api *ObjectAPI) Stat(ctx context.Context, path coreiface.Path) (*coreiface.ObjectStat, error) {
+	nd, err := api.core().ResolveNode(ctx, path)
+	if err != nil {
+		return nil, err
+	}
+
+	stat, err := nd.Stat()
+	if err != nil {
+		return nil, err
+	}
+
+	out := &coreiface.ObjectStat{
+		Cid:            nd.Cid(),
+		NumLinks:       stat.NumLinks,
+		BlockSize:      stat.BlockSize,
+		LinksSize:      stat.LinksSize,
+		DataSize:       stat.DataSize,
+		CumulativeSize: stat.CumulativeSize,
+	}
+
+	return out, nil
+}
+
+func (api *ObjectAPI) AddLink(ctx context.Context, base coreiface.Path, name string, child coreiface.Path, create bool) (coreiface.Node, error) {
+	rootNd, err := api.core().ResolveNode(ctx, base)
```

Comment:
Let's call these variables `base` instead of `root` -- the root of `/ipfs/QmFoo/a/b` would be QmFoo, but the base of this operation is `b`